### ISSUE TITLE
Redirect loop after update to 3.5.18

### ISSUE THREAD

**Issue by seaneble** [contao/core#8560]
We upgraded an installation from 3.5.17 to 3.5.18. The frontend is not reachable afterwards, accessing the home page returns code 302 redirecting to the same page over and over again.

Some analysis was possible before downgrading again (not too much because of the downtime).
The redirect is not related to any .htaccess magic (renamed for testing). The redirect comes from Contao (Cookie headers are also present in the HTTP response).

It does not seem to be related to #8547, because we don't have any PageRedirect pages in our tree.

It also does not seem to be related to #8552, because the error pages do not redirect in our setup, they are plain error pages with actual error messages.

Next step for us would be to try this in a testing environment. For the time being, I cannot recommend updating.

**Comment by seaneble** [comment]
It has to be something in f42f2a41b13a768fccb033f109c9e3eca8b86e47, but I was not involved in #8498 and don't know about the logic introduced there.

**Comment by fritzmg** [comment]
Can you make a copy of the installation available for debugging?

**Comment by fritzmg** [comment]
Btw. are you sure _all_ the redirects from the redirect loop are coming from Contao? Because you mentioned that the status code is `302`, however the status code of Contao's redirect implemented in https://github.com/contao/core/commit/f42f2a41b13a768fccb033f109c9e3eca8b86e47 is `301`.

**Comment by seaneble** [comment]
No, see line 362.

**Comment by fritzmg** [comment]
My bad, was looking at the wrong lines.

You can probably solve your infinite redirect loop by giving your start page the alias `index` - however, this of course not a solution for the underlying bug.

**Comment by leofeyer** [comment]
How do I reproduce the issue with a fresh installation and the Contao Official Demo?

**Comment by seaneble** [comment]
I'll provide a link to an SQL dump later in the day.

**Comment by fritzmg** [comment]
A complete copy would be needed, not just an SQL dump.

**Comment by seaneble** [comment]
Need more time, music_academy.sql is not a part of the LTS download archive anymore (which I expected).

**Comment by seaneble** [comment]
The problem only occurs because the page alias of our home page is `/`.
Currently I'm not sure why the alias is like that. The other question is whether it is something Contao should support.

**Comment by seaneble** [comment]
One approach of course would be to change the alias.
But on the other hand one could argue that the change introduced in Contao 3.5.18 is a BC break and using / as an alias should still be possible (regardless of the sense of it).

**Comment by fritzmg** [comment]
Having an alias like `/` does not make much sense indeed. If you want to enforce a trailing slash for the empty request, you should define a redirect in your web server's configuration or the `.htaccess` for that.

Regarding fixing it: I am not sure if it makes a lot of sense to accommodate such a special case.

**Comment by fritzmg** [comment]
Having an alias like `/` does not make much sense indeed. If you want to enforce a trailing slash for the empty request, you should define a redirect in your web server's configuration or the `.htaccess` for that.

Regarding fixing it: I am not sure if it makes a lot of sense to accommodate such a special case.

**Comment by seaneble** [comment]
I agree with the first part. But BC breaks are not nice and might happen to other people as well. It as allowed behaviour, it should stay allowed behaviour.

**Comment by TorCoGit** [comment]
I had the same problem today. Deleting the following line from my personally adapted .htaccess solved the problem:
`RewriteRule ^home\.html$ / [R=301,L]`

**Comment by leofeyer** [comment]
@seaneble Can you please post your `.htaccess` adjustments?

**Comment by seaneble** [comment]
```
273,274c273
<   RewriteCond %{REQUEST_FILENAME} !-d
<   RewriteRule .* index.php [L]
---
>   RewriteRule .*\.html$ index.php [L]
```

**Comment by leofeyer** [comment]
So your alias is `/.html`???

**Comment by seaneble** [comment]
Was. The diff is the other way around. Alias is empty.

**Comment by fritzmg** [comment]
@leofeyer seaneble's issue is fixed now anyway. The problem was, that he used `/` as the alias for the start page.

**Comment by leofeyer** [comment]
Using `/` as alias does not make sense IMHO. It might have been possible before Contao 3.5.18, but it was never explicitly supported.

**Comment by seaneble** [comment]
I agree with the fact that '/' never was a good idea. Nevertheless, people (at least we) did use it in some projects. The complete site was broken after a maintenance update, which should not happen.

Perhaps one could include a migration step to change the alias from / to empty in Contao 3.5.19?

**Comment by leofeyer** [comment]
I'm still trying to reproduce your setup, which I cannot as long as you are not providing the necessary details (such as the **current** `.htaccess` configuration instead of the **former** one).

**Comment by fritzmg** [comment]
The reproduction steps are:

1. Use an empty HTML suffix in the system configuration.
2. Set the alias of the start page to `/`.
3. Use the following directives in your `.htaccess` (regarding the URL suffix,)

```
RewriteCond %{REQUEST_FILENAME} !\.(htm|php|js|css|map|htc|png|gif|jpe?g|ico|xml|csv|txt|swf|flv|mp4|webm|ogv|mp3|ogg|oga|eot|otf|tt[cf]|woff2?|svgz?|pdf|zip|gz)$
RewriteCond %{REQUEST_FILENAME} !-f
RewriteCond %{REQUEST_FILENAME} !-d
RewriteRule .* index.php [L]
```

Then open `http://example.org` or `http://example.org/`. There should be infinite redirects.

**Comment by leofeyer** [comment]
Thank you @fritzmg, I was able to reproduce the issue now.

This is actually a quirk of the `Environment` class, which returns all paths relative to the `<base href="">` URL. Therefore `http://example.org/` becomes `''` in the `Environment` class, whereas it becomes `'/'` in the `$_SERVER['REQUEST_URI']` variable. At least this is the behavior with Nginx and PHP-FPM.

I could hardcode a fix for this specific case, but I don't know if I should. @contao/developers /cc

**Comment by seaneble** [comment]
I was going to provide a zip of the Music Academy with broken settings today, but thanks to @fritzmg that is superfluous now. Great!

**Comment by Whissi** [comment]
Today I run into the same problem after upgrading Contao. Please implement the mentioned fix at least for 3.5.x because it is a very bad experience when updating a LTS version and you notice that your start page doesn't work anymore.

**Comment by leofeyer** [comment]
Fixed in b565df6234473ab10eaa2fc67418924be5d9d475.
